L1Trigger/L1TTrackMatch/plugins/L1TrackVertexAssociationProducer.cc

```diff
@@ -149,6 +150,75 @@ class L1TrackVertexAssociationProducer : public edm::global::EDProducer<> {
     std::vector<double> deltaZMaxEtaBounds_;
     std::vector<double> deltaZMax_;
   };
+
+  struct NNTrackWordSelector {
+    NNTrackWordSelector(tensorflow::Session* AssociationSesh,
+                        const double AssociationThreshold,
+                        const std::vector<double>& AssociationNetworkZ0binning,
+                        const std::vector<double>& AssociationNetworkEtaBounds,
+                        const std::vector<double>& AssociationNetworkZ0ResBins)
+        : AssociationSesh_(AssociationSesh),
+          AssociationThreshold_(AssociationThreshold),
+          z0_binning_(AssociationNetworkZ0binning),
+          eta_bins_(AssociationNetworkEtaBounds),
+          res_bins_(AssociationNetworkZ0ResBins) {}
+
+    bool operator()(const TTTrackType& t, const l1t::VertexWord& v) const {
+      tensorflow::Tensor inputAssoc(tensorflow::DT_FLOAT, {1, 4});
+      std::vector<tensorflow::Tensor> outputAssoc;
+
+      TTTrack_TrackWord::tanl_t etaEmulationBits = t.getTanlWord();
+      ap_fixed<16, 3> etaEmulation;
+      etaEmulation.V = (etaEmulationBits.range());
+
+      auto lower = std::lower_bound(eta_bins_.begin(), eta_bins_.end(), etaEmulation.to_double());
+
+      int resbin = std::distance(eta_bins_.begin(), lower);
+      float binWidth = z0_binning_[2];
+      // Calculate integer dZ from track z0 and vertex z0 (use floating point version and convert internally allowing use of both emulator and simulator vertex and track)
+      float dZ =
+          abs(floor(((t.getZ0() + z0_binning_[1]) / (binWidth))) - floor(((v.z0() + z0_binning_[1]) / (binWidth))));
+
+      ap_uint<14> ptEmulationBits = t.getTrackWord()(TTTrack_TrackWord::TrackBitLocations::kRinvMSB - 1,
+                                                     TTTrack_TrackWord::TrackBitLocations::kRinvLSB);
+      ap_ufixed<14, 9> ptEmulation;
+      ptEmulation.V = (ptEmulationBits.range());
+
+      ap_ufixed<22, 9> ptEmulation_rescale;
+      ptEmulation_rescale = ptEmulation.to_double();
+
+      ap_ufixed<22, 9> resBinEmulation_rescale;
+      resBinEmulation_rescale = res_bins_[resbin];
+
+      ap_ufixed<22, 9> MVAEmulation_rescale;
+      MVAEmulation_rescale = t.getMVAQualityBits();
+
+      ap_ufixed<22, 9> dZEmulation_rescale;
```

Comment:
Can/should `14` and `22` be slightly more explained named constants?

Reply:
Unfortunately I cannot make 14 and 22 as ap_ufixed type needs to be defined at compile time, these constants come from the quantisation used by the neural network so we can add comments to explain where they come from?

Reply:
@Chriisbrown I believe a `constexpr` can be used as a compile time constant but the comment would be good enough.